### PR TITLE
aggregation should return a default values when aggregation is keyles…

### LINKED ISSUE #1455: count() returns zero rows but should return one row with value 0
Running this query in neo4j playground and RedisInsight gives diffrent results:

```
MATCH (a) WHERE a.nonExistentProperty = 123 RETURN count(a)
```

neo4j playground:
```
count(a)
--
1 | 0

```

RedisInsight:

```
INFO
No records and no modifications
```

According to this post I think a row with a value of zero should be returned for this query: 
https://neo4j.com/developer/kb/understanding-aggregations-on-zero-rows/

> Operators in Cypher produce rows, and they also execute per row. When there are no more rows (due to MATCH not finding matches, or WHERE clause filtering), then there is nothing left to execute upon, and remaining operations become no-op.
> One big exception to this are aggregation functions, most often count() and collect().

So I guess this could be  a bug in RedisGraph?

### PR DESCRIPTION
**Pull request by swilly22**
…s and no data was processed
Resolves #1455
I'm pretty sure there are a number of duplicates to this issue, @jeffreylovitz can you do a quick search and relate all relevant issues?

### CODE REVIEW

File: tests/flow/test_aggregation.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                           0,    # sum
+                           None, # avg
+                           0,    # stDev
+                           0,    # stDevP
+                           []    # collect
+                           ]
+
+        query = """MATCH (n) WHERE n.v = 'noneExisting'
+                   RETURN count(n), min(n), max(n), sum(n.v), avg(n.v),
+                   stDev(n.v), stDevP(n.v), collect(n)"""
```
Comment (AviAvni) [inline tests/flow/test_aggregation.py:31]: add test for percentileDisc, percentileCont

**Comment by jeffreylovitz** [comment]
Also resolves #2138 and #2091.

**Review by jeffreylovitz — CHANGES_REQUESTED** [review]
Besides the leak, I only see one potential problem here.
Default values aren't used when an `ORDER BY` is specified, because the ordering expressions will be used as group keys.

Neo4j has an error condition we don't have that blocks this in the majority of cases:
```
In a WITH/RETURN with DISTINCT or an aggregation, it is not possible to access variables declared before the WITH/RETURN: a (line 1, column 63 (offset: 62))
"MATCH (a) WHERE a.nonexistent = true RETURN COUNT(a) ORDER BY a.v"
```
We still have the misbehavior on strange but valid queries like:
```
GRAPH.QUERY G "MATCH (a) WHERE a.v = 5 RETURN COUNT(a) ORDER BY rand()"
1) 1) "COUNT(a)"
2) (empty array)
```
But this seems pretty pedantic; I don't know if it's worth addressing.

File: src/arithmetic/func_desc.h [hunk 2]
```diff
@@ -23,21 +23,50 @@ typedef void (*AR_Func_Free)(void *ctx);
...
+	const char *name;           // function name
+	SIValue default_value;      // aggregation default value
+	AggCBs agg_callbacks;       // aggregation
 } AR_FuncDesc;
 
-AR_FuncDesc *AR_FuncDescNew(const char *name, AR_Func func, uint min_argc, uint max_argc,
-							SIType *types, bool reducible, bool aggregate);
+typedef struct {
+
+} AR_AggFuncDesc;
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/arithmetic/func_desc.h:47]: need to wait for @swilly22 he partially add it and I don't know if he want to make a change

File: src/arithmetic/func_desc.c [hunk 3]
```diff
@@ -22,22 +22,49 @@ AR_FuncDesc *AR_FuncDescNew
...
+	desc->types                   =  types;
+	desc->privdata                =  NULL;
+	desc->min_argc                =  min_argc;
+	desc->max_argc                =  max_argc;
+	desc->aggregate               =  true;
+	desc->reducible               =  reducible;
+	desc->default_value           =  SI_NullVal(); 
+	desc->agg_callbacks.bfree     =  NULL;
+	desc->agg_callbacks.bclone    =  NULL;
+	desc->agg_callbacks.finalize  =  NULL;
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/arithmetic/func_desc.c:67]: see comment above

File: src/arithmetic/aggregate_funcs/agg_funcs.c [hunk 4]
```diff
@@ -478,11 +443,13 @@ void Register_AggFuncs() {
 	types = array_new(SIType, 2);
 	array_append(types, SI_ALL);
 	array_append(types, T_PTR);
-	func_desc = AR_FuncDescNew("collect", AGG_COLLECT, 2, 2, types, false, true);
+	func_desc = AR_AggFuncDescNew("collect", AGG_COLLECT, 2, 2, types, false);
 	AR_SetPrivateDataRoutines(func_desc, Aggregate_Free, Aggregate_Clone);
+	AR_SetDefaultValue(func_desc, SI_Array(0));
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/arithmetic/aggregate_funcs/agg_funcs.c:448]: yes I know I'll ask @swilly22 about this

File: src/arithmetic/aggregate_funcs/agg_funcs.c [hunk 5]
```diff
@@ -132,7 +132,7 @@ void Aggregate_Free
 
 	SIValue_Free(ctx->result);
 
-	if(ctx->private_data) agg_func->callbacks.free(ctx->private_data);
+	if(ctx->private_data && agg_func->callbacks.free) agg_func->callbacks.free(ctx->private_data);
```
Comment (swilly22) [inline src/arithmetic/aggregate_funcs/agg_funcs.c:135]: aggregation functions must have both `private_data` and a `free` callback, maybe we should replace this with an assertion

File: src/arithmetic/placeholder_funcs/placeholder_funcs.c [hunk 6]
```diff
@@ -8,7 +8,7 @@
 #include "../func_desc.h"
 #include "../../util/arr.h"
 
-SIValue AR_PATH_FILTER(SIValue *argv, int argc) {
+SIValue AR_PATH_FILTER(SIValue *argv, int argc, void *private_data) {
```
Comment (swilly22) [inline src/arithmetic/placeholder_funcs/placeholder_funcs.c:11]: is this function still necessary ?

File: src/arithmetic/arithmetic_expression.c [hunk 7]
```diff
@@ -145,6 +141,7 @@ static AR_ExpNode *_AR_EXP_CloneOp(AR_ExpNode *exp) {
 	const char *func_name = exp->op.f->name;
 	uint child_count = exp->op.child_count;
 	AR_ExpNode *clone = AR_EXP_NewOpNode(func_name, child_count);
+	if(clone->op.f->callbacks.clone) clone->op.private_data = clone->op.f->callbacks.clone(exp->op.private_data);
```
Comment (swilly22) [inline src/arithmetic/arithmetic_expression.c:144]:
```c
clone = clone->op.f->callbacks.clone;
void *pdata = exp->op.private_data;
if(pdata != NULL && clone != NULL) {
    // clone callback specified, use it to duplicate function's private data
    clone->op.private_data = clone(exp->op.private_data);
}
```

File: src/arithmetic/arithmetic_expression.c [hunk 8]
```diff
@@ -364,21 +360,18 @@ static bool _AR_EXP_ValidateInvocation
 	SIType actual_type;
 	SIType expected_type = T_NULL;
 
-	// if this is an aggregation function, reduce all user-facing counts by 1
-	int offset = fdesc->aggregate ? 1 : 0;
```
Comment (swilly22) [inline src/arithmetic/arithmetic_expression.c]: Nice!

**Review by swilly22 — COMMENTED** [review]
@AviAvni looking really good, two comments and we're ready for merge.

**Comment by swilly22** [comment]
We still need to look into @jeffreylovitz raised point:

> Default values aren't used when an ORDER BY is specified, because the ordering expressions will be used as group keys.
> Neo4j has an error condition we don't have that blocks this in the majority of cases:
> 
> In a WITH/RETURN with DISTINCT or an aggregation, it is not possible to access variables declared before the WITH/RETURN: a (line 1, column 63 (offset: 62))
> "MATCH (a) WHERE a.nonexistent = true RETURN COUNT(a) ORDER BY a.v"

File: src/arithmetic/placeholder_funcs/placeholder_funcs.c (hunk 6 above)
Comment (AviAvni) [reply src/arithmetic/placeholder_funcs/placeholder_funcs.c:11]: yes
in src/ast/ast_build_filter_tree.c _convertPatternPath

File: src/arithmetic/arithmetic_expression.c (hunk 7 above)
Comment (AviAvni) [reply src/arithmetic/arithmetic_expression.c:144]: in case of distinct we can't check if pdata is not null

**Review by AviAvni — APPROVED** [review]

**Pull request merged**